host/sgx/exception_sim.c

```diff
@@ -0,0 +1,449 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openenclave/host.h>
+#include <openenclave/internal/registers.h>
+#include <openenclave/internal/sgx/td.h>
+#include <signal.h>
+#if defined(_WIN32)
+#else
+#include <ucontext.h>
+#endif
+#include "enclave.h"
+#include "exception.h"
+
+#if defined(_WIN32)
+bool is_simulation(struct _EXCEPTION_POINTERS* exception_pointers)
+{
+    PCONTEXT context = exception_pointers->ContextRecord;
+    uint64_t tcs_address = context->Rbx;
+    oe_enclave_t* enclave = oe_query_enclave_instance((void*)tcs_address);
+    return enclave->simulate;
+}
+
+static void _oe_aex_sim(PCONTEXT context, void* host_fs)
+{
+    // Update cssa as AEX does in real mode.
+    sgx_tcs_t* tcs = (sgx_tcs_t*)(context->Rbx);
+    tcs->cssa++;
+
+    // This aex is delayed. Change the FS register to host side although this
+    // code is already on host side.
+    oe_set_fs_register_base(host_fs);
+}
+
+static sgx_ssa_gpr_t* _get_ssa_gpr(sgx_tcs_t* tcs)
+{
+    // why ossa != OE_SSA_FROM_TCS_BYTE_OFFSET?
+    uint32_t cssa = tcs->cssa;
+    // uint64_t ossa = ((sgx_tcs_t*)tcs_address)->ossa;
+
+    // oe_sgx_td_t* td = (oe_sgx_td_t*)(tcs_address + TD_FROM_TCS);
+    uint64_t ssa_frame_size = 0;
+    //    uint64_t ssa_frame_size = td->base.__ssa_frame_size;
+    if (ssa_frame_size == 0)
+    {
+        ssa_frame_size = OE_DEFAULT_SSA_FRAME_SIZE;
+    }
+
+    uint64_t ssa_base_address = (uint64_t)tcs + OE_SSA_FROM_TCS_BYTE_OFFSET;
+
+    // cssa always points to the unfilled ssa.
+    return (
+        sgx_ssa_gpr_t*)(ssa_base_address + cssa * ssa_frame_size * OE_PAGE_SIZE - OE_SGX_GPR_BYTE_SIZE);
+}
+
+static void _update_ssa_from_context(PCONTEXT context)
+{
+    sgx_tcs_t* tcs = (sgx_tcs_t*)(context->Rbx);
+    sgx_ssa_gpr_t* ssa_gpr = _get_ssa_gpr(tcs);
+
+    // Update gpr.
+    ssa_gpr->rax = context->Rax;
+    ssa_gpr->rbx = context->Rbx;
+    ssa_gpr->rcx = context->Rcx;
+    ssa_gpr->rdx = context->Rdx;
+    ssa_gpr->rsp = context->Rsp;
+    ssa_gpr->rbp = context->Rbp;
+    ssa_gpr->rsi = context->Rsi;
+    ssa_gpr->rdi = context->Rdi;
+    ssa_gpr->r8 = context->R8;
+    ssa_gpr->r9 = context->R9;
+    ssa_gpr->r10 = context->R10;
+    ssa_gpr->r11 = context->R11;
+    ssa_gpr->r12 = context->R12;
+    ssa_gpr->r13 = context->R13;
+    ssa_gpr->r14 = context->R14;
+    ssa_gpr->r15 = context->R15;
+    ssa_gpr->rip = context->Rip;
+    ssa_gpr->rflags = context->EFlags;
+
+    // This flag is checked by virtual dispacher.
+    ssa_gpr->exit_info.as_fields.valid = true;
+}
+
+static void _update_sgx_vector(PCONTEXT context, PEXCEPTION_RECORD exceptionrecord)
+{
+    // Hardcode here must match g_vector_to_exception_code_mapping[] in
+    // enclave/core/sgx/exception.c
+    uint32_t sgx_vector = (uint32_t)-1;
```

Comment:
For readability, consider defining the `OE_EXCEPTION_UNKNOWN` in this file and add comments that it was defined the same as in exception.h.

Reply:
Fixed, I did not notice there is OE_EXCEPTION_UNKNOWN. Thanks!